src/modules/sdbus/sdbus.c

```diff
@@ -0,0 +1,457 @@
+    }
+out:
+    return;
+log_drop:
+    flux_log (ctx->h, LOG_DEBUG, "bus drop %s", sdmsg_typestr (m));
+}
+
+static int handle_call_request (struct sdbus_ctx *ctx,
+                                const flux_msg_t *msg,
+                                flux_error_t *error)
+{
+    sd_bus_message *m;
+    json_t *req;
+    int e;
+    uint64_t cookie;
+
+    if (flux_request_unpack (msg, NULL, "o", &req) < 0) {
+        errprintf (error, "unable to decode call request");
+        return -1;
+    }
+    if (!(m = interface_request_fromjson (ctx->bus, req, error))) {
+        errno = EINVAL;
+        goto error;
+    }
+    if ((e = sd_bus_send (NULL, m, &cookie)) < 0) {
+        errno = -e;
+        errprintf (error, "error sending sdbus request: %s", strerror (errno));
+        goto error;
+    }
+
+    flux_log (ctx->h,
+              LOG_DEBUG,
+              "bus send %s cookie=%ju %s",
+              sdmsg_typestr (m),
+              (uintmax_t)cookie,
+              sd_bus_message_get_member (m));
+
+    if (flux_msg_aux_set (msg,
+                          "sdreq",
+                          m,
+                          (flux_free_f)sd_bus_message_unref) < 0) {
+        errprintf (error, "error saving call request state");
+        goto error;
+    }
+    return 0;
+error:
+    ERRNO_SAFE_WRAP (sd_bus_message_unref, m);
+    return -1;
+}
+
+static void call_cb (flux_t *h,
+                     flux_msg_handler_t *mh,
+                     const flux_msg_t *msg,
+                     void *arg)
+{
+    struct sdbus_ctx *ctx = arg;
+    flux_error_t error;
+    const char *errmsg = NULL;
+
+    if (flux_request_decode (msg, NULL, NULL) < 0)
+        goto error;
+    if (handle_call_request (ctx, msg, &error) < 0) {
+        errmsg = error.text;
+        goto error;
+    }
+    if (flux_msglist_append (ctx->requests, msg) < 0)
+        goto error;
+    return;
+error:
+    if (flux_respond_error (h, msg, errno, errmsg) < 0)
+        flux_log_error (h, "error responding to call request");
+}
+
+static void subscribe_continuation (flux_future_t *f, void *arg)
+{
+    struct sdbus_ctx *ctx = arg;
+
+    if (flux_rpc_get (f, NULL) < 0) {
+        flux_log (ctx->h,
+                  LOG_ERR,
+                  "subscribe failed: %s",
+                  future_strerror (f, errno));
+        bulk_respond_error (ctx->h,
+                            ctx->subscribers,
+                            errno,
+                            future_strerror (f, errno));
+    }
+}
+
+static void unsubscribe_continuation (flux_future_t *f, void *arg)
+{
+    struct sdbus_ctx *ctx = arg;
+
+    if (flux_rpc_get (f, NULL) < 0) {
+        flux_log (ctx->h,
+                  LOG_ERR,
+                  "unsubscribe failed: %s",
+                  future_strerror (f, errno));
+    }
+}
+
+/* Subscribe when first subscriber is added.
+ * Unsubscribe when last subscriber is removed.
+ */
+static void update_subscribe (struct sdbus_ctx *ctx, int delta)
```

Comment:
medium level comment: this function took me awhile to grok, perhaps just split up into subscribe and unsubscribe separate functions.

Reply:
Good idea, that whole thing got me confused when I looked at it again also!  I'll refactor some way that makes that clearer.